doc/fluid/design/inference_engine.md

```diff
@@ -0,0 +1,252 @@
+
+To help convert input/output data format between any Fluid operators and some `EngineOp`, a pair of `EngineInputConvertOp` and `EngineOutputConvertOp` needs to insert into the Fluid description. The reason for these converters is an operator, not a method are as follows
+
+- the converter works between the Fluid operators and `EngineOp`s, both of their data formats might be different and need to specify, for example
+  - `RNNOp -> xEngineOp`, the input is from an `LoDTensor` to `xTensor`
+  - `MulOp -> xEngineOp`, the input is from an `Tensor` to `xTensor`
+  - but `RNNOp -> MulOp -> xEngineOp`, the input is from an `LoDTensor` to `xTensor`
+  - the `EngineOp` can not get the external operators those link to it (the `RNNOp` and `MulOp` above), so it is impossible to deduce the data's format interact with the external Fluid framework.
+- the converter will result in additional overhead, to make it an operator is more clear and more flexible for further optimization.
+
+## Engine-related Design
+
+###  EngineOp                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
+
+`EngineOp` is just a normal Fluid operator, which has an attribute called `subblock` to get the Fluid description about a sub-block.
+
+To unify the behavior of different `EngineOp`s, an underlying interface over all the engines is raised as follows 
+
+```c++
+/*
+EngineBase is used like
+class xEngine : EngineBase { ... }
+
+xEngine engine;
+engine.Build(some_desc);
+
+for (const auto& batch : dataset) {
+    auto& input0 = engine.buffer("input0");
+    if (input0.device == DeviceType::CPU) {
+    std::memcpy(input0.buffer, batch.data(), batch.size());
+    input0.size = batch.size();
+    } else if (input0.device == DeviceType::GPU) {
+    	cudaMemCpy(input0.buffer, batch.data(), batch.size());
+        input0.size = batch.size();
+    }
+    engine.Execute();
+    
+    const auto& output0 = engine.buffer("output0");
+    // ...
+}
+*/
+enum class DeviceType {
+    CPU = 0,
+    GPU
```

Comment:
GPU=1?

Reply:
The enum syntax just needs to set the first element, and following elements will increase automatically.